### PR TITLE
NIR parser

### PR DESCRIPTION
**Pull request by Duhemm**
NIR parser using fastparse.

TODOS:
- [x] Correctly parse strings
  - [x] Normal strings
  - [x] Triple-quoted strings
- [x] Correctly parse IDs
- [ ] More testing

### CODE REVIEW

File: tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+      })
+    def Select =
+      P("select" ~ Val.parser ~ "," ~ Val.parser ~ "," ~ Val.parser map {
+        case (cond, thenp, elsep) => nir.Op.Select(cond, thenp, elsep)
+      })
+    def Classalloc = P("classalloc" ~ Global.parser map (nir.Op.Classalloc(_)))
+    def Field =
+      P("field[" ~ Type.parser ~ "]" ~ Val.parser ~ "," ~ Global.parser map {
+        case (ty, value, name) => nir.Op.Field(ty, value, name)
+      })
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala]: We actually never use the type in `Op.Field`. Lets remove it as a field altogether.

File: tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+    def Method =
+      P("method[" ~ Type.parser ~ "]" ~ Val.parser ~ "," ~ Global.parser map {
+        case (ty, value, name) => nir.Op.Method(ty, value, name)
+      })
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala]: The same here, we don't use type field, lets remove it.

**Comment by densh** [comment]
This is great, we'll finally be able to unit-test compiler passes in isolation now!

File: tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala (hunk 1 above)
Comment (densh) [inline tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala]: Yeah, lets change right here.

File: tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+trait NirParser[T] {
+  def parser: P[T]
+
+  final def apply(nir: String) =
+    parser.parse(nir)
+}
+
+object Base {
+  import fastparse.all._
+  def byte: P[Byte]   = P("-".!.? ~ CharIn('0' to '9').rep(1).! map { case (None, n) => n.toByte; case (_, n) => ("-" + n).toByte })
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala]: If I'm not mistaken, for fastparse it's idiomatic to have parsers defined as vals whenever possible, not defs. This ensures we don't instantiate them over and over again for no good reason.

Comment (Duhemm) [inline tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala]: Done

File: tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala (hunk 2 above)
Comment (Duhemm) [inline tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala]: Done

File: tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala (hunk 1 above)
Comment (Duhemm) [inline tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala]: Done

File: nir/src/main/scala/scala/scalanative/nir/Versions.scala [hunk 4]
```diff
@@ -20,8 +20,8 @@ object Versions {
    * when 1.3-based release happens all of the code needs to recompiled with
    * new version of the toolchain.
    */
-  final val compat: Int   = 6
-  final val revision: Int = 7
+  final val compat: Int   = 8
+  final val revision: Int = 9
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/Versions.scala]: `compat = 7`, `revision = 8`.

File: tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+      P("array" ~ Type.parser ~ "{" ~ Val.parser.rep(sep = ",") ~ "}" map {
+        case (ty, values) => nir.Val.Array(ty, values)
+      })
+    val Chars = P("c" ~ stringLit map (nir.Val.Chars(_)))
+    val Local =
+      P(NirParser.Local.parser ~ ":" ~ Type.parser map {
+        case (name, ty) => nir.Val.Local(name, ty)
+      })
+    val Global =
+      P(NirParser.Global.parser map (nir.Val.Global(_, nir.Type.None))) /// FIXME
```
Comment (Duhemm) [inline tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala]: @densh The type of a `Val.Global` doesn't appear in the HNIR. Should I add it?

File: tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala [hunk 6]
```diff
@@ -0,0 +1,466 @@
+package scala.scalanative
+package nir
+package parser
+
+import fastparse.WhitespaceApi
+import fastparse.noApi._
+
+trait NirParser[T] {
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala]: `Base[T]` and put both `Base` trait and companion in a separate file.

File: tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala (hunk 5 above)
Comment (densh) [inline tools/src/main/scala/scala/scalanative/nir/parser/NirParser.scala]: Nice catch, we should include it. NIR is supposed to be 100% explicit.

File: tools/src/test/scala/scala/scalanative/nir/parser/NirParserTest.scala [hunk 7]
```diff
@@ -0,0 +1,708 @@
...
+package parser
+
+import util.sh
+import Shows._
+
+import scala.io.Source
+import fastparse.all.Parsed
+import org.scalatest._
+
+class NirParserTest extends FlatSpec with Matchers {
```
Comment (densh) [inline tools/src/test/scala/scala/scalanative/nir/parser/NirParserTest.scala]: Similarly to splitting parsers apart, lets have one test suite per hierarchy.

**Comment by densh** [comment]
I don't think that we should include `hnir` file in the repo. It will screw up github statistics for good. Also given that it's derivative of linking a specific project, we should rather compute it on the fly.

**Comment by densh** [comment]
@Duhemm is there anything else you want to do here? Current state of things looks good to me.

**Comment by Duhemm** [comment]
I still want to add more testing, but I think this will come in another PR. Thanks for the review. Do you want me to squash all the commits together?

**Comment by densh** [comment]
Yeah, lets squash and merge this!

**Comment by Duhemm** [comment]
Squashed!

**Comment by densh** [comment]
Needs a rebase due to merge of GVN.

**Comment by Duhemm** [comment]
Done

**Comment by densh** [comment]
We forgot to add `tools/test` to the travis build sequence.

**Comment by Duhemm** [comment]
Good catch. I added it.

**Comment by densh** [comment]
LGTM, thanks!

**Pull request merged**